File: medium/Rotate_List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if head is None:
            return head
        while True:
            move2front = head
            head_copy = head
            count_k = 0
            node = ListNode(head_copy.val)
            node_next = node
            while head_copy.next != None:
                if count_k == k:
                    move2front = move2front.next
                    node_next.next = ListNode(move2front.val)
                    node_next = node_next.next
                else:
                    count_k+=1
                head_copy = head_copy.next
            if k-count_k > 0:
                k=k%(count_k+1)
            else:
                break
        head_copy.next = node    
        return move2front.next

File: cpp/medium/Rotate_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (head == nullptr) return head;
        int length = 1;
        ListNode* next_head = head;
        ListNode* answer;
        while(next_head->next != nullptr) {
            ++length;
            next_head = next_head->next;
        }
        next_head->next = head;
        while(k >= length) k-=length;
        for(int i = 0; i <= length-k-1; i++) {
            next_head = next_head->next;
        }
        answer = next_head->next;
        next_head->next = nullptr;
        return answer;
    }
};
